src/django_components/finders.py

```diff
@@ -0,0 +1,144 @@
+import os
+from fnmatch import fnmatch
+from typing import Any, Dict, Iterable, List, Optional, Tuple, Union
+
+from django.contrib.staticfiles.apps import StaticFilesConfig
+from django.contrib.staticfiles.finders import BaseFinder
+from django.contrib.staticfiles.utils import get_files
+from django.core.checks import CheckMessage, Error, Warning
+from django.core.files.storage import FileSystemStorage
+from django.utils._os import safe_join
+
+from django_components.app_settings import app_settings
+from django_components.template_loader import get_dirs
+
+# To keep track on which directories the finder has searched the static files.
+searched_locations = []
+
+
+# Custom Finder for staticfiles that searches for all files within the directories
+# defined by `COMPONENTS.dirs`.
+#
+# This is what makes it possible to define JS and CSS files in the directories as
+# defined by `COMPONENTS.dirs`, but still use the JS / CSS files with `static()` or
+# `collectstatic` command.
+class ComponentsFileSystemFinder(BaseFinder):
+    """
+    A static files finder based on ``FileSystemFinder``.
+
+    Differences:
+    - This finder uses ``COMPONENTS.dirs`` setting to locate files instead of ``STATICFILES_DIRS``.
+    - This finder ignores Python and HTML files (``.py``, ``.pyc``, ``.html``, ``.django``)
+    - If ``COMPONENTS.dirs`` is not set or empty, defaults to ``settings.BASE_DIR / "components"``
+    """
+
+    IGNORE_PATTERNS = StaticFilesConfig.ignore_patterns + ["*.html", "*.django", "*.py", "*.pyc"]
```

Comment:
I'm a little worried about the security implications of this. If someone accidentally would put some other file type in these dirs, they could get exposed on the internet.

At the same time, explicitly including file types is hard, because the full list of potential static files is very large...

Any ideas of how to make this both safe and future-proof?

Reply:
Maybe a combo of both?

```python
ALLOWED_EXTENSIONS = getattr(settings, 'COMPONENTS_ALLOWED_EXTENSIONS', 
                             ['.css', '.js', '.png', '.jpg', '.jpeg', '.gif', '.svg', '.woff', '.woff2', '.ttf', '.eot'])
FORBIDDEN_EXTENSIONS = ['.py', '.pyc', '.html', '.django', '.env', '.db', '.sqlite3']
```

Reply:
I didn't bother to change this, but agree with you, explicitly listing allowed extensions is better than the opposite. The reason why I used only ignored patterns was because then we can delegate the filtering to [`django.contrib.staticfiles.utils.get_files`](https://github.com/django/django/blob/main/django/contrib/staticfiles/utils.py#L16), which acepts a list of glob patterns.

But globbing doesn't support negations, so one cannot search for "everything except `.html`". So we can't use `get_files` to get only files with allowed patterns. And so we need to do the filtering ourselves.

And TBH, if we're doing the filtering ourselves, I would go with regex over globs, so users can be more expressive, e.g. maybe they would want to exclude CSS only from certain components, or omit static files that start with underscore.

But I just tried it, and writing regex for patterns like "everything except `.html`, `.py`, or `.pyc`" is not very easy to read. So I would still go with as you suggested, specifying both allowed and forbidden patterns. One approach I've seen that I liked was that first all files from allowed patterns are found, and then any forbidden files are removed. So, in other words, if a file was matched successfully against both `ALLOWED_EXTENSIONS` and `FORBIDDEN_EXTENSIONS`, then the file will be OMITTED.

```py
COMPONENTS = {
	"allowed_static_files": [
		r"\.{css|js|png|jpg|jpeg|svg}$",
	],
	"forbidden_static_files": [
		r"\.{html|py|pyc}$",
	],
}
```

The defaults would then be:
```py
COMPONENTS = {
	"allowed_static_files": [
		# Match all files
		".*",
	],
	"forbidden_static_files": [
		# Exclude .html, .py, .pyc
		r"\.{html|py|pyc}$",
	],
}
```